pkg/runs/runsDownload.go

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright contributors to the Galasa project
+ */
+package runs
+
+import (
+	"bufio"
+	"context"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/galasa.dev/cli/pkg/api"
+	galasaErrors "github.com/galasa.dev/cli/pkg/errors"
+	"github.com/galasa.dev/cli/pkg/galasaapi"
+	"github.com/galasa.dev/cli/pkg/utils"
+)
+
+// DownloadArtifacts - performs all the logic to implement the `galasactl runs download` command,
+// but in a unit-testable manner.
+func DownloadArtifacts(
+	runName string,
+	forceDownload bool,
+	fileSystem utils.FileSystem,
+	timeService utils.TimeService,
+	console utils.Console,
+	apiServerUrl string) error {
+
+	var err error = nil
+	var runs []galasaapi.Run
+	var artifactPaths []string
+
+	runs, err = GetRunsFromRestApi(runName, timeService, apiServerUrl)
+	if err == nil {
+		for _, run := range runs {
+			runId := run.GetRunId()
+			artifactPaths, err = GetArtifactPathsFromRestApi(runId, apiServerUrl)
+			if err == nil {
+				for _, artifactPath := range artifactPaths {
+					artifactData, err := GetFileFromRestApi(runId, strings.TrimPrefix(artifactPath, "/"), apiServerUrl)
+					if err != nil {
+						return err
+					} else {
+						err = WriteArtifactToFileSystem(fileSystem, runName, artifactPath, artifactData, forceDownload)
+						if err != nil {
+							return err
+						}
+					}
+				}
+			}
+		}
+	}
+	return err
+}
+
+// Retrieves the paths of all artifacts for a given test run using its runId.
+func GetArtifactPathsFromRestApi(runId string, apiServerUrl string) ([]string, error) {
+
+	var err error = nil
+	var artifactPaths []string
+	log.Println("Retrieving artifact paths for the given run")
+
+	// An HTTP client which can communicate with the api server in an ecosystem.
+	restClient := api.InitialiseAPI(apiServerUrl)
+
+	var artifactsList []galasaapi.ArtifactIndexEntry
+	artifactsList, httpResponse, err := restClient.ResultArchiveStoreAPIApi.
+		GetRasRunArtifactList(context.Background(), runId).
+		Execute()
+
+	if err != nil {
+		err = galasaErrors.NewGalasaError(galasaErrors.GALASA_ERROR_RETRIEVING_ARTIFACTS_FAILED, err.Error())
+	} else {
+		for _, artifact := range artifactsList {
+			artifactPaths = append(artifactPaths, artifact.GetPath())
+		}
+	}
+	defer httpResponse.Body.Close()
+	log.Printf("%v artifact path(s) found", len(artifactPaths))
+
+	return artifactPaths, err
+}
+
+// Writes an artifact to the host's file system, creating a new directory for the run's artifacts
+// to be written to. Existing files are only overwritten if the --force option is used as part of
+// the "runs download" command.
+func WriteArtifactToFileSystem(
+	fileSystem utils.FileSystem,
+	runDirectory string,
+	artifactPath string,
+	fileDownloaded *os.File,
```

Comment:
I've always tried to not have os.* in any code so far, as it's hard to mock out...
Shouldn't it be some kind of Mock file so we can set the contents in tests without actually touching the file system ?

Reply:
Maybe using io.Writer instead of os.File ? Then we can implement a mock writer easily enough ?

Reply:
Addressed in https://github.com/galasa-dev/cli/pull/92/commits/b6a487aea49009dc2efbec738888c6add97648af - FileMock.go already implements the io.Writer interface and is used to mock out downloaded files in the tests.